File: 13.roman-to-integer.ts
/*
 * @lc app=leetcode id=13 lang=typescript
 *
 * [13] Roman to Integer
 */

// @lc code=start

const hash = {
    I: 1,
    V: 5,
    X: 10,
    L: 50,
    C: 100,
    D: 500,
    M: 1000,
};

function romanToInt(s: string): number {
    let ans = hash[s[s.length - 1]];

    for (let i = 0; i < s.length - 1; i++) {
        if (hash[s[i]] < hash[s[i + 1]]) ans -= hash[s[i]];
        else ans += hash[s[i]];
    }

    return ans;
}
// @lc code=end

File: 13.roman-to-integer.cs
/*
 * @lc app=leetcode id=13 lang=csharp
 *
 * [13] Roman to Integer
 */

// @lc code=start
public class Solution
{
    public int RomanToInt(string s)
    {
        var roman = new Dictionary<char, int>() {
            {'M', 1000},
            {'D', 500},
            {'C', 100},
            {'L', 50},
            {'X', 10},
            {'V', 5},
            {'I', 1},
        };
        int z = 0;
        for (int i = 0; i < s.Length - 1; i++)
        {
            if (roman[s[i]] < roman[s[i + 1]])
            {
                z -= roman[s[i]];
            }
            else
            {
                z += roman[s[i]];
            }
        }
        return z + roman[s[s.Length - 1]];
    }
}
// @lc code=end

